File: cpp/src/Mechanical/SensoredGearbox.cpp
//#include "ctre/phoenix/Mechanical/SensoredGearbox.h"
//
//using namespace CTRE::MotorControl;
//
//namespace CTRE{ namespace Mechanical{
//
//SensoredGearbox::SensoredGearbox(float gearRatio, SmartMotorController *talon, SmartMotorController::FeedbackDevice feedbackDevice){
//	master = talon;
//	_gearRatio = gearRatio;
//	master->SetFeedbackDevice(feedbackDevice);
//}
//SensoredGearbox::SensoredGearbox(float gearRatio, SmartMotorController *talon, IFollower *slave1, SmartMotorController::FeedbackDevice feedbackDevice){
//	master = talon;
//	_gearRatio = gearRatio;
//	master->SetFeedbackDevice(feedbackDevice);
//	slave1->Follow(master);
//}
//SensoredGearbox::SensoredGearbox(float gearRatio, SmartMotorController *talon, IFollower *slave1, IFollower *slave2, SmartMotorController::FeedbackDevice feedbackDevice){
//	master = talon;
//	_gearRatio = gearRatio;
//	master->SetFeedbackDevice(feedbackDevice);
//	slave1->Follow(master);
//	slave2->Follow(master);
//}
//SensoredGearbox::SensoredGearbox(float gearRatio, SmartMotorController *talon, IFollower *slave1, IFollower *slave2, IFollower *slave3, SmartMotorController::FeedbackDevice feedbackDevice){
//	master = talon;
//	_gearRatio = gearRatio;
//	master->SetFeedbackDevice(feedbackDevice);
//	slave1->Follow(master);
//	slave2->Follow(master);
//	slave3->Follow(master);
//}
//float SensoredGearbox::GetPosition(){
//	return master->GetPosition() / _gearRatio;
//}
//float SensoredGearbox::GetVelocity(){
//	return master->GetSpeed() / _gearRatio;
//}
//void SensoredGearbox::SetCurrentLimit(unsigned int currentLimitAmps){
//	master->SetCurrentLimit(currentLimitAmps);
//}
//void SensoredGearbox::Set(float output){
//	if(_isInverted)
//		output = -output;
//	master->Set(output);
//}
//void SensoredGearbox::SetControlMode(CTRE::MotorControl::ControlMode::SmartControlMode mode){
//	master->SetControlMode(mode);
//}
//void SensoredGearbox::SetLimits(float forwardLimit, float reverseLimit){
//	master->ConfigForwardLimit(forwardLimit);
//	master->ConfigReverseLimit(reverseLimit);
//}
//void SensoredGearbox::DisableLimits(){
//	master->ConfigLimitMode(2);
//}
//void SensoredGearbox::SetInverted(bool invert){
//	_isInverted = invert;
//}
//bool SensoredGearbox::GetInverted() const{
//	return _isInverted;
//}
//void SensoredGearbox::InvertSensor(bool invert){
//	master->SetSensorDirection(invert);
//}
//void SensoredGearbox::SetSensor(float position){
//	master->SetPosition(position);
//}
//void SensoredGearbox::ConfigNominalOutputVoltage(float forwardVoltage, float reverseVoltage){
//	master->ConfigNominalOutputVoltage(forwardVoltage, reverseVoltage);
//}
//void SensoredGearbox::ConfigPeakOutputVoltage(float forwardVoltage, float reverseVoltage){
//	master->ConfigPeakOutputVoltage(forwardVoltage, reverseVoltage);
//}
//void SensoredGearbox::SetVoltageCompensationRampRate(float rampRate){
//	master->SetVoltageCompensationRampRate(rampRate);
//}
//void SensoredGearbox::SetVoltageRampRate(float rampRate){
//	master->SetVoltageRampRate(rampRate);
//}
//void SensoredGearbox::SetMotionMagicCruiseVelocity(float RPM){
//	master->SetMotionMagicCruiseVelocity(RPM);
//}
//void SensoredGearbox::SetMotionMagicAcceleration(float RPM){
//	master->SetMotionMagicAcceleration(RPM);
//}
//CTRE::MotorControl::IMotorController *SensoredGearbox::GetMaster(){
//	return master;
//}
//
//
//}}

File: java/src/com/ctre/phoenix/mechanical/SensoredGearbox.java
//package com.ctre.phoenix.mechanical;
//
//import com.ctre.phoenix.motorcontrol.can.BaseMotorController;
//
//public class SensoredGearbox //implements com.ctre.phoenix.Signals.IInvertable
//{
//	private double _unitsPerRevolution;
//	private double _scalar;
//	
//	private boolean _isInverted;
//	public BaseMotorController _controller;
//	
//	
//	public SensoredGearbox(double unitsPerRevolution, doublBaseMotorController mc1, BaseMotorController.FeedbackDevice feedback)
//	{
//		_controller = mc1;
//		_unitsPerRevolution = unitsPerRevolution;
//		_controller.setFeedbackDevice(feedback);
//	}
//	public SensoredGearbox(double unitsPerRevolution, BaseMotorController mc1, IFollower mc2, BaseMotorController.FeedbackDevice feedback)
//	{
//		_controller = mc1;
//		_unitsPerRevolution = unitsPerRevolution;
//		_controller.setFeedbackDevice(feedback);
//		
//		mc2.follow(_controller);
//	}
//	public SensoredGearbox(double unitsPerRevolution, BaseMotorController mc1, IFollower mc2, IFollower mc3, BaseMotorController.FeedbackDevice feedback)
//	{
//		_controller = mc1;
//		_unitsPerRevolution = unitsPerRevolution;
//		_controller.setFeedbackDevice(feedback);
//		
//		mc2.follow(_controller);
//		mc3.follow(_controller);
//	}
//	public SensoredGearbox(double unitsPerRevolution, BaseMotorController mc1, IFollower mc2, IFollower mc3, IFollower mc4, BaseMotorController.FeedbackDevice feedback)
//	{
//		_controller = mc1;
//		_unitsPerRevolution = unitsPerRevolution;
//		_controller.setFeedbackDevice(feedback);
//
//		mc2.follow(_controller);
//		mc3.follow(_controller);
//		mc4.follow(_controller);
//	}
//	
//	 /**
//     * @return The geared output position in rotations.
//     */
//    public double getPosition()
//    {
//        return (double)com.ctre.phoenix.Util.scaleNativeUnitsToRotations(_unitsPerRevolution, (long)_controller.getPosition());
//    }
//    /**
//     * @return The geared output velocity in rotations per minute.
//     */
//    public double getVelocity()
//    {
//        return (double)com.ctre.phoenix.Util.scaleNativeUnitsToRpm(_unitsPerRevolution, (long)_controller.getSpeed());
//    }
//
//    public void SetCurrentLimit(int currentLimitAmps, int timeoutms)
//    {
//        _controller.setCurrentLimit(currentLimitAmps);
//    }
//
//    public void set(double output)
//    {
//        if (_isInverted)
//            output = -output;
//        
//        _controller.set(output);
//    }
//
//    /** Set the control mode of the gearbox */
//    public void SetControlMode(ControlMode.SmartControlMode mode)
//    {
//        _controller.setControlMode(mode);
//    }
//
//    /** Set the limits on the forward and reverse drive */
//    public void SetLimits(double forwardLimit, double reverseLimit)
//    {
//        _controller.setForwardSoftLimit(com.ctre.phoenix.Util.scaleRotationsToNativeUnits(_unitsPerRevolution, forwardLimit));
//        _controller.setReverseSoftLimit(com.ctre.phoenix.Util.scaleRotationsToNativeUnits(_unitsPerRevolution, reverseLimit));
//    }
//    
//    /** Disable limits of the gearbox */
//    public void disableLimits()
//    {
//        _controller.enableLimitSwitch(false, false);
//    }
//
//    /* IInvertable */
//    public void setInverted(boolean invert)
//    {
//        _isInverted = invert;
//    }
//
//    public boolean getInverted()
//    {
//        return _isInverted;
//    }
//
//    public void invertSensor(boolean invert)
//    {
//        _controller.reverseSensor(invert);
//    }
//
//    public void setSensor(double position)
//    {
//        _controller.setPosition(com.ctre.phoenix.Util.scaleRotationsToNativeUnits(_unitsPerRevolution, position));
//    }
//
//    public void configNominalOutputVoltage(double forwardVoltage, double reverseVoltage)
//    {
//        _controller.configNominalOutputVoltage(forwardVoltage, reverseVoltage);
//    }
//
//    public void configPeakOutputVoltage(double forwardVoltage, double reverseVoltage)
//    {
//        _controller.configPeakOutputVoltage(forwardVoltage, reverseVoltage);
//    }
//
//    public void setVoltageCompensationRampRate(double RampRate)
//    {
//        _controller.setVoltageCompensationRate(RampRate);
//    }
//
//    public void setVoltageRampRate(double RampRate)
//    {
//        _controller.setVoltageRampRate(RampRate);
//    }
//
//    /* IMotionMagical */
//    /* Motion Magic stuff */
//    public void setMotionMagicCruiseVelocity(double RPM)
//    {
//        _controller.setMotionMagicCruiseVelocity(com.ctre.phoenix.Util.scaleVelocityToNativeUnits(_unitsPerRevolution, RPM));
//    }
//
//    public void setMotionMagicAcceleration(double RPM)
//    {
//        _controller.setMotionMagicAcceleration(com.ctre.phoenix.Util.scaleVelocityToNativeUnits(_unitsPerRevolution, RPM));
//    }
//
//    public BaseMotorController getMaster()
//    {
//        return _controller;
//    }
//}*/